enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,327 @@
+
+- Provide certificates for `127.0.0.1`, `localhost`, or NodeIP service endpoints
+
+  (See discussion in “Implementation Details/Notes/Constraints”)
+
+- Provide individual certificates that uniquely identify individual pods in the StatefulSet,
+  protecting the intra-StatefulSet communication against re-routing to unintended members
+  of that StatefulSet
+  (or to a different StatefulSet attached to the same headless service).
+
+## Proposal
+
+Currently, a service annotated with
+a `service.beta.openshift.io/serving-cert-secret-name: $secretName`
+requests the `service-ca` controller to generate a secret `$secretName`
+containing a certificate+key pair for `${service.name}.${service.namespace}.svc`
+(and a `….cluster.local`) as subjects.
+
+For headless services, generate certificates that include two more subjects,
+`*.${service.name}.${service.namespace}.svc` and
+`*.${service.name}.${service.namespace}.svc.cluster.local`,
+intended to match `${pod.hostname}.${service.name}.${service.namespace}.svc…`
+served in DNS for endpoints of that service
+(and in particular, for StatefulSets,
+`${statefulset.name}-${podID}.${service.name}.${service.namespace}.svc…`).
+
+
+### User Stories
+
+#### Creating a TLS-protected StatefulSet
+
+**Note:** This only ensures that the communication happens between members of the StatefulSet,
+it does not provide individual identity.
+
+- The user creates/annotates a headless service with
+  `service.beta.openshift.io/serving-cert-secret-name: $secretName`
+- The Service CA generates a `$secretName` secret, with type: `kubernetes.io/tls`,
+  containing a single certificate and private key for
+  - `${service.name}.${service.namespace}.svc`
+  - `${service.name}.${service.namespace}.svc.cluster.local`
+  - `*.${service.name}.${service.namespace}.svc`
+  - `*.${service.name}.${service.namespace}.svc.cluster.local`
+- The user creates/updates a StatefulSet for the application to contain use the generated secret
+
+#### Scaling the StatefulSet Up/Down
+
+Once the secret is generated, users can scale the StatefulSet with no change to the usual process.
+
+#### Using TLS to Guarantee Individual Identity of StatefulSet Pods, or StatefulSets
+
+To ensure that a compromised StatefulSet pod can’t impersonate other pods from the
+same StatefulSet,
+or any other pods attached ot the same headless service
+(if this goal is achievable for the application,
+i.e. if the pods treat the intra-StatefulSet communication as untrusted
+and can plausibly protect against a rogue StatefulSet member):
+
+- Use something else, not the certificates generated by the Service CA.
+- **Don’t accept the Service CA as a valid root of trust for connections intended
+  for individual StatefulSet pods**.
+
+#### Certificate Expiration, Manual Secret Regeneration, CA Rotations
+
+No change to existing Service CA use.
+
+### Implementation Details/Notes/Constraints
+
+#### Not Generating Certificates for `127.0.0.1`, `localhost`, or NodeIPs
```

Comment:
I’m afraid I still need to follow up on that, it’s on my list.

In general, this is OK _only_ if the CA is completely single-purpose (only signs certificates for a single service, and clients only trust it for connection to that single service), i.e. no other client could ever be misled by an attacker presenting such a `localhost`/IP certificate.

In other cases, it would be better for the CA to only sign a subject that identifies the service (in this case I guess the `etcd.*` host names). That does not prevent connections from bypassing DNS or the external network and going to 127.0.0.1 — the clients can be configured to make a TCP connection to one host name and to verify the certificate for another, probably by setting `TLSClientConfig.ServerName` (compare https://github.com/golang/go/issues/5829 ).